lib/magma/retrieval.rb

```diff
@@ -10,23 +10,25 @@ def initialize(model, record_names, attribute_names, opts={})
         page_size: MAX_PAGE_SIZE,
         page: 1
       }.merge(opts)
+
+      @filters = opts[:filters] || []
+      @collapse_tables = opts[:collapse_tables]
+      @page_size = opts[:page_size]
+      @page = opts[:page]
+      @restrict = opts[:restrict]
+
       @model = model
       @record_names = record_names
 
       # the retrieval filters out attributes that are not allowed
       @requested_attribute_names = attribute_names
       @attribute_names = attributes.map(&:name)
-
-      @filters = opts[:filters] || []
-      @collapse_tables = opts[:collapse_tables]
-      @page_size = opts[:page_size]
-      @page = opts[:page]
     end
 
     def attributes
       @attributes ||= begin
         attributes = @model.attributes.values.select do |att|
-          !att.is_a?(Magma::TableAttribute) && requested?(att)
+          !att.is_a?(Magma::TableAttribute) && requested?(att) && !restricted?(att)
```

Comment:
This could be confusing. Attributes that don't exist and attributes that are restricted are handled the same. You might want to return some meta data to the users to specify attributes that don't but are requested, attributes you are restricted from, and whether the query is run as a restricted user.

Reply:
The user can ascertain what attributes are available, restricted and not from the json template, which is maybe adequate documentation. I'm not sure about adding rejected arguments to the return value - it might be useful to raise an argument error if bad or restricted attributes are requested instead